File: src/main/kotlin/guitarproblem/Inventory.java
package guitarproblem;


import guitarproblem.enums.Builder;
import guitarproblem.enums.Type;
import guitarproblem.enums.Wood;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class Inventory {
    private List guitars;

    public Inventory() {
        guitars = new LinkedList();
    }

    public void addGuitar(String serialNumber, double price,
                          Builder builder, String model,
                          Type type, int numString, Wood backWood, Wood topWood) {
        GuitarSpec guitarSpec = new GuitarSpec( builder,
                model, type,numString, backWood, topWood);
        Guitar guitar = new Guitar(serialNumber, price,guitarSpec);
        guitars.add(guitar);
    }

    public Guitar getGuitar(String serialNumber) {
        for (Iterator i = guitars.iterator(); i.hasNext(); ) {
            Guitar guitar = (Guitar) i.next();
            if (guitar.getSerialNumber().equals(serialNumber)) {
                return guitar;
            }
        }
        return null;
    }

    public List search(GuitarSpec searchSpec) {
        List matchingGuitars = new LinkedList();
        for (Iterator i = guitars.iterator(); i.hasNext(); ) {
            Guitar guitar = (Guitar)i.next();
            if (guitar.getGuitarSpec().matches(searchSpec))
                matchingGuitars.add(guitar);
        }
        return matchingGuitars;
    }
}

File: src/main/kotlin/guitarproblem/Inventory.kt
package guitarproblem

class Inventory {
     private val inventory: MutableList<Instrument> = mutableListOf()

    fun addInstrument(
        serialNumber: String, price: Double, instrumentSpec: InstrumentSpec
    ) {
        var instrument: Instrument? = null
        if (instrumentSpec is GuitarSpec) {
            instrument = Guitar(serialNumber, price, instrumentSpec)
        } else if (instrumentSpec is MandolinSpec) {
            instrument = Mandolin(serialNumber, price, instrumentSpec)
        }
        inventory.add(instrument!!)
    }

    //    fun getInstrument(serialNumber: String): Guitar? {
//        for (guitar in Instrument) {
//            if (guitar.serialNumber == serialNumber) {
//                return guitar
//            }
//        }
//        return null
//    }
    fun get(serialNumber: String): Instrument? {

        for (instrument in inventory) {
            if (instrument.serialNumber == serialNumber) {
                return instrument
            }
        }
        return null
    }

    fun search(searchSpec: InstrumentSpec): List<Instrument> {
        val matchingInstrument: MutableList<Instrument> = mutableListOf()
        for (instrument in inventory) {
            if (instrument.spec.matches(searchSpec)) matchingInstrument.add(instrument)
        }
        return matchingInstrument
    }
}